include/deal.II/particles/property_pool.h

```diff
@@ -115,14 +124,37 @@ namespace Particles
      */
     const unsigned int n_properties;
 
+    /**
+     * The currently allocated properties (whether assigned to
+     * a particle or available for assignment).
+     */
+    std::vector<double> properties;
+
     /**
      * A collection of handles that have been created by
-     * allocate_properties_array() and have not been destroyed by
-     * deallocate_properties_array().
+     * allocate_properties_array() and have been destroyed by
+     * deallocate_properties_array(). Since the memory is still
+     * allocated these handles can be reused for new particles
+     * to avoid memory allocation.
      */
-    std::set<Handle> currently_open_handles;
+    std::vector<Handle> currently_available_handles;
   };
 
+    /* ---------------------- inline and template functions ------------------ */
+
+  inline ArrayView<double>
+  PropertyPool::get_properties(const Handle handle)
+  {
+    const std::vector<double>::size_type data_index =
+      (handle != invalid_handle) ? handle * n_properties : 0;
+    Assert(data_index <= properties.size() - n_properties,
+           ExcMessage("Invalid property handle. This can happen if the "
+                      "handle was duplicated and then one copy was deallocated "
+                      "before trying to access the properties."));
```

Comment:
Shouldn't we, at least in principle, also check that the handle hasn't been deallocated? I reckon that that might be expensive. If you decide that that shouldn't happen, then maybe leaving a comment here would be nice about why not.

Reply:
I left a comment here:
```
    // Ideally we would need to assert that 'handle' has not been deallocated
    // by searching through 'currently_available_handles'. However, this
    // is expensive and this function is performance critical, so instead
    // just check against the properties range, and rely on the fact
    // that handles are invalidated when handed over to
    // deallocate_properties_array().
```

`get_properties` is likely the function that gets called most often in this class, and we can not search through the vector (or whatever other container) of open handles here, unless the access is possible in constant time and very fast. Invalidating handles when deallocating them should help preventing access.